File: playground/gemini-tts.ts
// Gemini 2.5 Pro TTS - text-to-speech with bilingual support
import { execSync } from "child_process";
import { writeFileSync } from "fs";

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
const text = `Read this text naturally, speaking Vietnamese and Chinese words in their original languages:
Chào bạn! Chào mừng đến với bài học 'Câu chuyện về lòng tử tế'. Trong buổi đầu tiên này, chúng ta sẽ bắt đầu với tình huống khẩn cấp khi nhân vật chính nhận tin con bị bệnh nặng (严重). Anh ấy phải vay tiền và vội vã đi trong đêm. Hãy cùng học những từ vựng đầu tiên để hiểu rõ hoàn cảnh này nhé! Bắt đầu thôi (开始)!`;

async function main() {
  const res = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro-preview-tts:generateContent?key=${GEMINI_API_KEY}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: [{ parts: [{ text }] }],
        generationConfig: {
          responseModalities: ["AUDIO"],
          speechConfig: { voiceConfig: { prebuiltVoiceConfig: { voiceName: "Kore" } } },
        },
      }),
    }
  );

  const json = await res.json();
  if (json.error) return console.error("Error:", json.error.message);

  const audioData = json.candidates?.[0]?.content?.parts?.[0]?.inlineData?.data;
  if (!audioData) return console.log("No audio data received");

  writeFileSync("gemini-output.pcm", Buffer.from(audioData, "base64"));
  execSync("ffmpeg -y -f s16le -ar 24000 -ac 1 -i gemini-output.pcm gemini-output.mp3 2>/dev/null");
  console.log("Audio saved to gemini-output.mp3");
}

main();

File: na/gemini-tts.py
from dotenv import load_dotenv
load_dotenv()
import os, sys, subprocess, tempfile, base64
from pathlib import Path
import requests

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")

def generate_tts(text: str, output_path: str):
    prompt = f"Read this text at normal speed, speaking Vietnamese, and English words in their original languages:\n{text}"

    res = requests.post(
        f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-pro-preview-tts:generateContent?key={GEMINI_API_KEY}",
        headers={"Content-Type": "application/json"},
        json={
            "contents": [{"parts": [{"text": prompt}]}],
            "generationConfig": {
                "responseModalities": ["AUDIO"],
                "speechConfig": {"voiceConfig": {"prebuiltVoiceConfig": {"voiceName": "Kore"}}},
            },
        },
    )

    data = res.json()
    if "error" in data:
        raise Exception(data["error"]["message"])

    audio_data = data.get("candidates", [{}])[0].get("content", {}).get("parts", [{}])[0].get("inlineData", {}).get("data")
    if not audio_data:
        raise Exception("No audio data received")

    tmp = tempfile.mktemp()
    pcm_path = f"{tmp}.pcm"
    with open(pcm_path, "wb") as f:
        f.write(base64.b64decode(audio_data))
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    subprocess.run(["ffmpeg", "-y", "-f", "s16le", "-ar", "24000", "-ac", "1", "-i", pcm_path, output_path], capture_output=True)
    os.unlink(pcm_path)
    print(f"Saved: {output_path}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python gemini-tts.py <input_file.txt>")
        sys.exit(1)
    input_file = Path(sys.argv[1])
    text = input_file.read_text()
    output_path = os.path.join("output", input_file.with_suffix(".mp3").name)
    generate_tts(text, output_path)
